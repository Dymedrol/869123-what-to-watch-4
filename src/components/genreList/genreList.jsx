import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {ActionCreator} from "./../../reducer.js";

class GenreList extends PureComponent {
  constructor(props) {
    super(props);

    this.getGenreList = this.getGenreList.bind(this);
  }

  getGenreList() {
    const genres = new Set();
    genres.add(`All genres`);
    this.props.allMovies.forEach((movie) => {
      genres.add(movie.genre);
    });
    const genreList = Array.from(genres);

    return genreList;
  }

  render() {
    const {onClick} = this.props;
    const genreList = this.getGenreList();
    const genre = this.props.genre;

    return <React.Fragment>
      {genreList.map((item) => {
        const activeClass = genre === item ? `catalog__genres-item--active` : ``;

        return <li
          key={item}
          className={`catalog__genres-item ${activeClass}`}
          onClick={() => {
            onClick(item);
          }}
        >
          <a href="#" className="catalog__genres-link">{item}</a>
        </li>;
      })}
    </React.Fragment>;
  }
}

const mapStateToProps = (state) => ({
  movies: state.movies,
  allMovies: state.allMovies,
  genre: state.genre,
});

const mapDispatchToProps = (dispatch) => ({
  onClick(genre) {
    dispatch(ActionCreator.changeGenre(genre));
    dispatch(ActionCreator.getFilmsByGenre());
  }
});

GenreList.propTypes = {
  genre: PropTypes.string.isRequired,
  allMovies: PropTypes.array.isRequired,
  onClick: PropTypes.func.isRequired,
  movies: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    src: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
  })),
};

export {GenreList};
export default connect(mapStateToProps, mapDispatchToProps)(GenreList);
