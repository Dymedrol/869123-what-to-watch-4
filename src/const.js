const Genres = {
  ALL: `All genres`,
  COMEDIES: `Comedies`,
  CRIME: `Crime`,
  DOCUMENTARY: `Documentary`,
  DRAMA: `Drama`,
  HORROR: `Horror`,
  KIDS_FAMILY: `Kids & Family`,
  ROMANCE: `Romance`,
  SCI_FI: `Sci-Fi`,
  THRILLERS: `Thrillers`,
};

const Tabs = [
  `Overview`,
  `Details`,
  `Reviews`
];

const MovieListStep = {
  MAIN: 8,
  MOVIEPAGE: 4,
};

const videoPlayerModes = {
  SMALL: `small`,
  FULLSCREEN: `fullscreen`,
};

export {Genres, Tabs, MovieListStep, videoPlayerModes};
