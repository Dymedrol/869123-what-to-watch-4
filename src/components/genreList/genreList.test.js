import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';

import {GenreList} from './genreList.jsx';
import {Genres} from '../../const.js';
import MOVIES from '../../mocks/films.js';

const mockStore = configureStore([]);

it(`render GenresList`, () => {

  const store = mockStore({
    genre: Genres.ALL,
    movies: MOVIES,
  });

  const genres = [`All genres`, `Crime`];

  const tree = renderer.create(
      <Provider store={store}>
        <GenreList
          genre={Genres.ALL}
          onClick={()=>{}}
          movies={MOVIES}
          allMovies={MOVIES}
          setActiveItem={()=>{}}
          genreList = {genres}
        />
      </Provider>, {
        createNodeMock: () => {
          return {};
        }})
    .toJSON();


  expect(tree).toMatchSnapshot();
});
