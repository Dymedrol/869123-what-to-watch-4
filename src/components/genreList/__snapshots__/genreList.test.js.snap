// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render GenresList 1`] = `
Array [
  <li
    className="catalog__genres-item catalog__genres-item--active"
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      All genres
    </a>
  </li>,
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      comedy
    </a>
  </li>,
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      drama
    </a>
  </li>,
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      science fiction
    </a>
  </li>,
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Crime
    </a>
  </li>,
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Documentary
    </a>
  </li>,
]
`;
